idb/idb.go

```diff
@@ -106,13 +106,44 @@ func DecodeTxnRowNext(s string) (round uint64, intra uint32, err error) {
 	return
 }
 
+// OptionalUint wraps bool and uint. It has a custom marshaller below.
+type OptionalUint struct {
+	Present bool
+	Value   uint
+}
+
+// MarshalText implements TextMarshaler interface.
+func (ou OptionalUint) MarshalText() ([]byte, error) {
+	if !ou.Present {
+		return nil, nil
+	}
+	return []byte(fmt.Sprintf("%d", ou.Value)), nil
+}
+
+// UnmarshalText implements TextUnmarshaler interface.
+func (ou *OptionalUint) UnmarshalText(text []byte) error {
+	if text == nil {
+		*ou = OptionalUint{}
+	} else {
+		value, err := strconv.ParseUint(string(text), 10, 64)
+		if err != nil {
+			return err
+		}
+		*ou = OptionalUint{
+			Present: true,
+			Value:   uint(value),
+		}
+	}
+
+	return nil
+}
+
 // TxnExtra is some additional metadata needed for a transaction.
 type TxnExtra struct {
 	AssetCloseAmount uint64 `codec:"aca,omitempty"`
-	// RootIntra is set on inner transactions. Combined with the confirmation
+	// RootIntra is set only on inner transactions. Combined with the confirmation
 	// round it can be used to lookup the root transaction.
-	// The type is string to allow distinguishing between 0 and empty.
-	RootIntra string `codec:"root-intra,omitempty"`
```

Comment:
It just occurred to me that we can use `RootTxid` to determine if this field exists or not. We can avoid special encoding/decode code entirely and use a uint64 type here and use `len(RootTxid) > 0` as the "Present" condition.

Reply:
The query would need to change though to handle the special case when `t.extra->>'root-intra'` is empty string. https://github.com/algorand/indexer/pull/721/files#diff-8fbec00be6e6b2f77a294cbe9a3ee8e3fbdcb5eb13be3046e5da8277302264c2R637
If you would like to change it, I will not object.

By the way, I'm afraid this query has quadratic complexity. I'm not sure postgres is smart enough to look up the right `root` txn with one query. Possibly, for each `txn` row there will be a scan of all rows for this round.